IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,471 @@
+                              throttleType:(MSIDThrottlingType)throttleType
+                               returnError:(NSError *_Nullable *_Nullable)error
+{
+    NSString *thumbprint = nil;
+    NSString *throttleTypeString = nil;
+    NSInteger throttleDuration = 0;
+    
+    NSDate *retryHeaderDate = [self getRetryDateFromErrorResponse:errorResponse];
+    
+    switch (throttleType)
+    {
+        case MSIDThrottlingType429:
+            thumbprint = request.strictRequestThumbprint;
+            throttleTypeString = @"MSIDThrottlingType429";
+            if (!retryHeaderDate)
+            {
+                throttleDuration = Default429Throttling;
+            }
+            else
+            {
+                NSTimeInterval MAX_THROTTLING_TIME = MaxRetryAfter;
+                NSDate *max429ThrottlingDate = [[NSDate date] dateByAddingTimeInterval:MAX_THROTTLING_TIME];
+                NSTimeInterval timeDiff = [retryHeaderDate timeIntervalSinceDate:max429ThrottlingDate];
+                throttleDuration = (timeDiff > MAX_THROTTLING_TIME) ? MAX_THROTTLING_TIME : timeDiff;
+            }
+            break;
+        case MSIDThrottlingTypeUIRequired:
+            thumbprint = request.fullRequestThumbprint;
+            throttleTypeString = @"MSIDThrottlingTypeUIRequired";
+            throttleDuration = DefaultUIRequired;
+            break;
+        default:
+            break;
+    }
+    
+    MSIDThrottlingCacheRecord *record = [[MSIDThrottlingCacheRecord alloc] initWithErrorResponse:errorResponse
+                                                                                    throttleType:throttleTypeString
+                                                                                throttleDuration:throttleDuration];
+    [self.cacheService setObject:record forKey:thumbprint error:error];        
+}
+
+/**
+// TODO:  Huge TODO Here
+ */
+- (void)updateServerTelemetry:(MSIDThrottlingCacheRecord *)cacheRecord
+{
+    
+}
+
+- (NSDate *)getRetryDateFromErrorResponse:(NSError * _Nullable)errorResponse
+{
+    NSDate *retryHeaderDate = nil;
+    NSString *retryHeaderString = nil;
+    NSDictionary *headerFields = errorResponse.userInfo[MSIDHTTPHeadersKey];
+    retryHeaderString = headerFields[@"Retry-After"];
+    retryHeaderDate = [NSDate msidDateFromRetryHeader:retryHeaderString];
+    return retryHeaderDate;
+}
+
++ (BOOL)validateInput:(id<MSIDThumbprintCalculatable> _Nonnull)request
+{
+    return (!request.fullRequestThumbprint && !request.strictRequestThumbprint);
+}
+
+/**
+ Get last refresh time from our NSUserDefaults.
+ */
++ (NSDate *)getLastRefreshTimeWithContext:(id<MSIDRequestContext>)context
+                                    error:(NSError*__nullable*__nullable)error
+{
+    NSDictionary *searchQuery = [self prepareSearchQuery];
+    CFDictionaryRef result = nil;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)searchQuery, (CFTypeRef *)&result);
```

Comment:
Is this needed to share refresh date between processes?
I think it's OK to have it per-process to reduce a performance hit that would come with read/write operations.